File: ruby/secret_handshake.rb
class SecretHandshake
  attr_reader :commands

  def initialize(num)
    @commands=to_binary(num)
  end

  def to_binary(num, handshake=[])
    return handshake if num.class != Fixnum
    [16,8,4,2,1].each_with_index do |i, index|
      if num-i >= 0
        handshake.unshift(ACTIONS[index])
        num-=i
        break if num==0
      end
    end
    handshake.last=="reverse" ? handshake[0...-1].reverse : handshake
  end

  ACTIONS=["reverse", "jump", "close your eyes", "double blink", "wink"]

end

File: js/secret-handshake.js
SecretHandshake = function(num){
  this.binary = toBinary(num)
};

function toBinary(num, digits=[]){
  if (!Number.isInteger(num)) {throw "Handshake must be a number"};
  [16,8,4,2,1].forEach(function(div){
    if (num-div>=0){
      digits.push(1);
      num-=div;
    }
    else{
      digits.push(0)
    };
  });
  return digits
};

SecretHandshake.prototype.commands = function(combo=[]){
  this.binary.reverse().forEach(function(digit, index){
    if (digit===1){
      index===4 ? combo.reverse() : combo.push(ACTIONS[index])
    }
  });
  return combo
};

const ACTIONS = ["wink", "double blink", "close your eyes", "jump"]

module.exports = SecretHandshake
